src/flamenco/repair/fd_recorder.c

```diff
@@ -0,0 +1,732 @@
+
+  recorder->high_priority_idx = 0;
+  recorder->medium_priority_idx = 0;
+  recorder->low_priority_idx = 0;
+  recorder->zero_hr_idx = 0;
+
+  fd_recorder_peer_map_t * peer_map = fd_recorder_peer_map( recorder );
+  fd_recorder_peer_t * peer_pool = fd_recorder_peer_pool( recorder );
+
+  /* First pass: count total peers and separate zero HR peers */
+  ulong non_zero_hr_peers = 0;
+
+  for( fd_recorder_peer_map_iter_t iter = fd_recorder_peer_map_iter_init( peer_map, peer_pool );
+       !fd_recorder_peer_map_iter_done( iter, peer_map, peer_pool );
+       iter = fd_recorder_peer_map_iter_next( iter, peer_map, peer_pool ) ) {
+
+    fd_recorder_peer_t * peer = fd_recorder_peer_map_iter_ele( iter, peer_map, peer_pool );
+
+    /* Separate zero hit rate peers immediately */
+    if( peer->ewma_hr <= 0.0 ) recorder->zero_hr_peers[recorder->zero_hr_cnt++] = &peer->key;
+    else                       non_zero_hr_peers++;
+  }
+
+  /* If no non-zero HR peers, we're done */
+  if( FD_UNLIKELY( non_zero_hr_peers == 0 ) ) return;
+
+  /* Allocate temporary array for sorting non-zero HR peers */
+  struct peer_rtt_pair sorted_peers[FD_MAX_PEERS];
+  ulong sorted_count = 0;
+
+  /* Second pass: collect non-zero HR peers for sorting */
+  for( fd_recorder_peer_map_iter_t iter = fd_recorder_peer_map_iter_init( peer_map, peer_pool );
+       !fd_recorder_peer_map_iter_done( iter, peer_map, peer_pool );
+       iter = fd_recorder_peer_map_iter_next( iter, peer_map, peer_pool ) ) {
+
+    fd_recorder_peer_t * peer = fd_recorder_peer_map_iter_ele( iter, peer_map, peer_pool );
+
+    /* Skip zero HR peers (already handled) */
+    if( FD_UNLIKELY( peer->ewma_hr <= 0.0 ) ) continue;
+
+    /* Add to sorting array */
+    sorted_peers[sorted_count].key = &peer->key;
+    sorted_peers[sorted_count].rtt = peer->ewma_rtt;
+    sorted_count++;
+  }
+
+  /* Sort peers by RTT (quickest first) */
+  qsort( sorted_peers, sorted_count, sizeof(struct peer_rtt_pair), peer_rtt_compare );
+
+  /* Distribute sorted peers into buckets:
+     - First 500: high priority
+     - Next 500: medium priority
+     - Remaining: low priority */
+
+  for( ulong i = 0; i < sorted_count; i++ ) {
+    if( i < 500 ) {
+      /* First 500 quickest -> high priority */
+      recorder->high_priority_peers[recorder->high_priority_cnt++] = sorted_peers[i].key;
+    } else if( i < 1000 ) {
+      /* Next 500 quickest -> medium priority */
+      recorder->medium_priority_peers[recorder->medium_priority_cnt++] = sorted_peers[i].key;
+    } else {
+      /* Remaining -> low priority */
+      recorder->low_priority_peers[recorder->low_priority_cnt++] = sorted_peers[i].key;
+    }
+  }
+}
+
+fd_pubkey_t *
+fd_recorder_select_peer(fd_recorder_t * recorder) {
+  if( FD_UNLIKELY( !recorder ) ) {
+    FD_LOG_WARNING(( "NULL recorder" ));
+    return NULL;
+  }
+
+  /* Check if we need to reshuffle */
+  if( FD_UNLIKELY( recorder->cycle_position == 0 && (recorder->cycle_count % 10000) == 0 ) ) {
```

Comment:
can you make this parameterizable? also see comment below on using an incremental data structure instead